src/python/foglamp/core/scheduler.py

```diff
@@ -547,60 +791,431 @@ def _schedule_next_task(self, schedule):
         await self._get_schedules()
 
     def _resume_check_schedules(self):
+        """Wakes up :meth:`_main_loop` so that
+        :meth:`_check_schedules` will be called the next time 'await'
+        is invoked.
+
+        """
         if self._main_sleep_task:
-            self._main_sleep_task.cancel()
+            try:
+                self._main_sleep_task.cancel()
+                self._main_sleep_task = None
+            except RuntimeError:
+                self._check_processes_pending = True
+        else:
+            self._check_processes_pending = True
 
     async def _main_loop(self):
-        """Main loop for the scheduler
-
-        - Reads configuration and schedules
-        - Runs :meth:`Scheduler._check_schedules` in an endless loop
-        """
+        """Main loop for the scheduler"""
         # TODO: log exception here or add an exception handler in asyncio
-        await self._mark_tasks_interrupted()
-        await self._read_storage()
 
         while True:
             next_start_time = await self._check_schedules()
 
             if self._paused:
                 break
 
-            if next_start_time is None:
+            # Determine how long to sleep
+            if self._check_processes_pending:
+                self._check_processes_pending = False
+                sleep_seconds = 0
+            elif next_start_time is None:
                 sleep_seconds = self._MAX_SLEEP
             else:
                 sleep_seconds = next_start_time - time.time()
 
-            self._logger.info("Sleeping for %s seconds", sleep_seconds)
+            if sleep_seconds > 0:
+                self._logger.info("Sleeping for %s seconds", sleep_seconds)
+                self._main_sleep_task = asyncio.ensure_future(asyncio.sleep(sleep_seconds))
+
+                try:
+                    await self._main_sleep_task
+                    self._main_sleep_task = None
+                except asyncio.CancelledError:
+                    self._logger.debug("Main loop awakened")
+
+    @staticmethod
+    async def populate_test_data():
```

Comment:
There are two reasons why not. First, the table metadata should not be duplicated. Secondly, database access will be moved to the storage layer.